### ISSUE TITLE
New cyclic dependency

### ISSUE THREAD

**Issue by pullmoll** [void-linux/void-packages#20438]
We have a new cyclic dependency: `pinentry` -> `gcr` -> `gnupg2` -> `pinentry`

The dependency is just a runtime dependency yet it prevents a bootstrapped arch (armv5tel here) from successfully breaking out of the cycle. Can we perhaps prevent runtime `depends=…` from creating such cycles?

**Comment by st3r4g** [comment]
I think `xbps-src` could be improved to handle such a scenario, which is not so uncommon. For example, `libglvnd` needs at least one GL vendor to work, but adding `depends=mesa` would introduce an endless loop at the moment.
One could detect the loop and break out of it, but it doesn't feel very elegant. Instead, what about making `xbps-src` build all the runtime dependencies **after** their "parent" package (i.e. the one that is being built)? I did a little experiment in this branch: https://github.com/st3r4g/void-packages/tree/xbps-src-rdeps
Obviously this is not feasible right now, as packages would be added to the index before some of their dependencies are built. But I believe this could work with a better staging mechanism, handled outside of `xbps-rindex`.
The current staging mechanism cannot handle multiple repositories, so it needs to be improved anyway. It was proposed something like: `xbps-rindex` always stages, a new tool (multirepo aware) checks for publishing the packages that are ready (shlibs and runtime deps checks).

**Comment by xtraeme** [comment]
A solution is to make a pinentry pkg that simply creates pinentry (main pkg, curses) and pinentry-tty. And then split the -qt and -gtk parts to another pkg.

**Comment by xtraeme** [comment]
Note that #20508 does not help in this case!

**Comment by xtraeme** [comment]
build cycles should not be workarounded, they must not exist at all.

Other build systems are full of stupid workarounds just for this. I refuse to introduce stuff like this into xbps-src.

**Comment by st3r4g** [comment]
but it's not a real build cycle... building all dependencies (runtime too) first is probably simpler but not the best solution imho. My proposal seems simple enough, why is it bad? Splitting packages is also a not very elegant workaround, they must be kept in sync manually and such...

**Comment by xtraeme** [comment]
@st3r4g I can't comment unless I see the code.

**Comment by st3r4g** [comment]
https://github.com/st3r4g/void-packages/commit/415b1820205ce7f38744ae81a68eb06a65ce5858

**Comment by xtraeme** [comment]
hmm just looked into it, seems a good idea after all!

so what your code does is just deferring building runtime deps until the pkg is built, isn't it?

**Comment by xtraeme** [comment]
@st3r4g but I'm not sure if this is enough.

**Comment by Chocimier** [comment]
@st3r4g: When dependencies are build after desired package is indexed, and some transitive dependency fail to build, one end up with non-installable package and no easy way to build just what is missing. Currently, it is enough to fix failing package and build desired package again.

Complete solution may need a temporary area of packages visible to xbps-src, but not indexed yet.

**Comment by st3r4g** [comment]
> @st3r4g: When dependencies are build after desired package is indexed, and some transitive dependency fail to build, one end up with non-installable package and no easy way to build just what is missing. Currently, it is enough to fix failing package and build desired package again.
> 
> Complete solution may need a temporary area of packages visible to xbps-src, but not indexed yet.

Exactly, the index must be updated only after all runtime dependencies are built successfully. In https://github.com/void-linux/xbps/issues/259 a new tool was proposed to handle proper indexing of packages (in a multi-repo aware way). So if such tool is made, we can postpone building runtime deps and let the new tool do the shlibs checks and rundeps checks, so that repo index has always installable packages.

**Comment by ericonr** [comment]
Still an issue, have just renamed.
